### PR TITLE
Refactored TypesFinder static methods into invokable classes

### PR DESCRIPTION
**Pull request by asgrim**

### CODE REVIEW

File: src/TypesFinder/FindParameterType.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * Given a function and parameter, attempt to find the type of the parameter
+     *
+     * @todo needs some context going on here...
+     * @param ReflectionFunctionAbstract $function
+     * @param ParamNode $node
+     * @return Type[]
+     */
+    public function __invoke(ReflectionFunctionAbstract $function, ParamNode $node)
```
Comment (Ocramius) [inline src/TypesFinder/FindParameterType.php]: To be tested

File: src/TypesFinder/FindPropertyType.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class FindPropertyType
+{
+    /**
+     * Given a property, attempt to find the type of the property
+     *
+     * @param PropertyNode $node
+     * @param ReflectionProperty $reflectionProperty
+     * @return Type[]
+     */
+    public function __invoke(PropertyNode $node, ReflectionProperty $reflectionProperty)
```
Comment (Ocramius) [inline src/TypesFinder/FindPropertyType.php]: To be tested

File: src/TypesFinder/FindTypeFromAst.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+class FindTypeFromAst
+{
+    /**
+     * Given an AST type, attempt to find a resolved type
+     *
+     * @todo resolve with context
+     * @param $astType
+     * @return \phpDocumentor\Reflection\Type|null
+     */
+    public function __invoke($astType)
```
Comment (Ocramius) [inline src/TypesFinder/FindTypeFromAst.php]: To be tested

File: src/TypesFinder/ResolveTypes.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+use phpDocumentor\Reflection\TypeResolver;
+
+class ResolveTypes
+{
+    /**
+     * @param string[] $stringTypes
+     * @param Context $context
+     * @return \phpDocumentor\Reflection\Type[]
+     */
+    public function __invoke($stringTypes, Context $context = null)
```
Comment (Ocramius) [inline src/TypesFinder/ResolveTypes.php]: To be tested

File: src/TypesFinder/FindParameterType.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        $docBlock = new DocBlock($function->getDocComment());
+
+        $paramTags = $docBlock->getTagsByName('param');
+
+        foreach ($paramTags as $paramTag) {
+            /* @var $paramTag \phpDocumentor\Reflection\DocBlock\Tag\ParamTag */
+            if ($paramTag->getVariableName() == '$' . $node->name) {
```
Comment (Ocramius) [inline src/TypesFinder/FindParameterType.php]: Strict comparisons

File: src/TypesFinder/FindPropertyType.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        $comment = $node->getAttribute('comments')[0];
+        $docBlock = new DocBlock(
+            $comment->getReformattedText(),
+            new DocBlock\Context(
+                $reflectionProperty->getDeclaringClass()->getNamespaceName()
+            )
+        );
+
+        /* @var \phpDocumentor\Reflection\DocBlock\Tag\VarTag $varTag */
+        $varTag = $docBlock->getTagsByName('var')[0];
```
Comment (Ocramius) [inline src/TypesFinder/FindPropertyType.php]: This assumes that a `@var` tag is present, and that it is the only `@var` tag.

File: src/TypesFinder/ResolveTypes.php (hunk 4 above)
Comment (Ocramius) [inline src/TypesFinder/ResolveTypes.php]: I would not make `Context $context` optional, as it leads to very subtle bugs

File: src/TypesFinder/FindTypeFromAst.php (hunk 3 above)
Comment (Ocramius) [inline src/TypesFinder/FindTypeFromAst.php]: No context passed in?

Comment (asgrim) [inline src/TypesFinder/FindTypeFromAst.php]: Issue to add context: #29

File: src/TypesFinder/ResolveTypes.php (hunk 4 above)
Comment (asgrim) [inline src/TypesFinder/ResolveTypes.php]: Once #29 and #30 are resolved, #31 can be done ;)

File: src/TypesFinder/FindParameterType.php (hunk 1 above)
Comment (asgrim) [inline src/TypesFinder/FindParameterType.php]: Tests added

File: src/TypesFinder/FindPropertyType.php (hunk 2 above)
Comment (asgrim) [inline src/TypesFinder/FindPropertyType.php]: same

**Pull request merged**